### PR TITLE
Fix-65575 MarkdownIt plugins (or Rules) called multiple times for one input

### LINKED ISSUE #65575: MarkdownIt plugins (or Rules) called multiple times for one input
- VSCode Version: Windows 7 & Mac 10.14 (Only these 2 Tested)

A user of my extension reports duplicate html elements in export markdown, I debug & find this issue as described in header. 

It doesn't affect the preview rendering, so I cannot tell the step to reproduce. But I can tell, it's reproducible when all extension (except mine) is disabled, and the issue is gone in extension debugging environment.

Please find more details here: https://github.com/qjebbs/vscode-markdown-extended/issues/28

PS: My extension uses VSCode Markdown render instance to provide WYSIWYG export

### PR DESCRIPTION
**Pull request by skprabhanjan**
@mjbvz , here is the initial PR based on my understanding to fix #65575 , I have added CachedToken and implementation to use it.
Please review this and let me know if anything is missing or not correct, I will correct it accordingly, Thanks :)

### CODE REVIEW

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for taking a look. I think this needs some more work though so that it won't cause regressions:

- The cache needs to use the document uri as the key for storing the tokens. One way to do this would be to create a new private method called `tokenize` or something like that that both `render` and `parse` call into. The `tokenize` method would check the cached tokens to see if they are valid for the current document. If they are, it would use these. Otherwise, it would compute and cache the tokens for the new document

- However there is also a small difference in `render` and `parse`: `parse` always strips out markdown yaml front-matter while `render` only strips the front-matter if a setting is configured. One way to preserve this behavior would be to run markdown-its `render` on the tokens without the front matter and then add the front matter back on to the html output

Let me know if that makes sense

**Comment by skprabhanjan** [comment]
@mjbvz , Cool , this sounds good and makes sense , I am working it and update you on the progress. Thanks for the review :)

**Comment by skprabhanjan** [comment]
@mjbvz , I have changed it as per your review , Please review it now and let me know if this is good or something needs to be changed , Happy to correct it (Just want to make sure it is good).
Thanks :)

**Review by mjbvz — CHANGES_REQUESTED** [review]
Looks light the right direction. Please take a look my comments and let me know if they make sense

File: extensions/markdown-language-features/src/markdownEngine.ts [hunk 1]
```diff
@@ -94,6 +97,23 @@ export class MarkdownEngine {
 		return { text, offset };
 	}
 
+	private tokenize(document: SkinnyTextDocument, engine?: MarkdownIt): Token[] {
```
[the remark this answers is not in the document]
Comment (skprabhanjan) [reply extensions/markdown-language-features/src/markdownEngine.ts]: I was thinking something else and that is not required, fixed it :)

File: extensions/markdown-language-features/src/markdownEngine.ts [hunk 2]
```diff
@@ -107,24 +127,25 @@ export class MarkdownEngine {
...
 		const engine = await this.getEngine(document.uri);
-
-		return engine.parse(text.replace(UNICODE_NEWLINE_REGEX, ''), {}).map(token => {
-			if (token.map) {
-				token.map[0] += offset;
-				token.map[1] += offset;
-			}
-			return token;
-		});
+		let tokens: Token[];
```
[the remark this answers is not in the document]
Comment (skprabhanjan) [reply extensions/markdown-language-features/src/markdownEngine.ts]: The tokenize method is clean now :)

**Comment by mjbvz** [comment]
Thanks! Will make a few tweaks to this but approach looks good

**Pull request merged**